### PR TITLE
fix: Properly handle selection on union types

### LINKED ISSUE #120: Incorrect __typename handling for union types
See [UnionsSpecificationTest](https://github.com/stuebingerb/KGraphQL/blob/main/kgraphql/src/test/kotlin/com/apurebase/kgraphql/specification/typesystem/UnionsSpecificationTest.kt#L87-L103):

```
@Test
fun `A Union type should allow requesting __typename`() {
    val result = execute(
        """{
        actors {
            name
            favourite {
                ... on Actor { name }
                ... on Director { name, age }
                ... on Scenario { content(uppercase: false) }
                __typename
            }
        }
    }""".trimIndent()
    )
    println(result)
}
```

The test passes (unsurprisingly) but does not actually contain a `__typename` in the response.

Expected result from https://graphql.com/learn/interfaces-and-unions/#querying-with-unions:

```
query GetFeaturedStallOrderOffers {
  featuredStall {
    orders {
      orderOffers {
        __typename
        ... on Discount {
          percent
        }
        ... on Coupon {
          amount
        }
        ... on ComplimentaryItem {
          limitPerCustomer
        }
        ... on Refund {
          isPartialAmount
        }
      }
    }
  }
}
```
```
{
  "data": {
    "featuredStall": {
      "orders": [
        {
          "orderOffers": [
            {
              "__typename": "Discount",
              "percent": 0.15
            }
          ]
        },
        {
          "orderOffers": [
            {
              "__typename": "Refund",
              "isPartialAmount": true
            }
          ]
        },
        {
          "orderOffers": [
            {
              "__typename": "ComplimentaryItem",
              "limitPerCustomer": 1
            }
          ]
        }
      ]
    }
  }
}
```

### PR DESCRIPTION
**Pull request by stuebingerb**
Resolves #120

### CODE REVIEW

File: kgraphql/src/main/kotlin/com/apurebase/kgraphql/schema/structure/RequestInterpreter.kt [hunk 1]
```diff
@@ -210,25 +210,53 @@ class RequestInterpreter(private val schemaModel: SchemaModel) {
...
+        //  do not define any fields, so *no* fields may be queried on this type without the use of type refining
+        //  fragments or inline fragments (with the exception of the meta-field `__typename`)."
         val unionMembersChildren: Map<Type, List<Execution>> =
             field.returnType.possibleTypes.associateWith { possibleType ->
                 val selections = selectionNode.selectionSet?.selections
+                // Add __typename to all fragment selections if specified
+                val typenameSelection = selections?.firstOrNull { it is FieldNode && it.name.value == "__typename" }
 
-                val a = selections?.filterIsInstance<FragmentSpreadNode>()?.firstOrNull {
+                val namedFragment = selections?.filterIsInstance<FragmentSpreadNode>()?.firstOrNull {
```
Comment (stuebingerb) [inline kgraphql/src/main/kotlin/com/apurebase/kgraphql/schema/structure/RequestInterpreter.kt]: Related: I believe multiple fragments on the same type _should_ be allowed and then merged in the result set. Will (have to) read up on that and adjust if needed.

File: kgraphql/src/main/kotlin/com/apurebase/kgraphql/schema/structure/RequestInterpreter.kt [hunk 2]
```diff
@@ -210,25 +210,53 @@ class RequestInterpreter(private val schemaModel: SchemaModel) {
...
+                // At this point only __typename was requested
+                if (typenameSelection != null) {
+                    return@associateWith handleReturnType(
+                        ctx,
+                        possibleType,
+                        selectionNode.selectionSet!!.addSelectionNode(typenameSelection)
+                    )
+                }
 
                 throw ValidationException(
```
Comment (stuebingerb) [inline kgraphql/src/main/kotlin/com/apurebase/kgraphql/schema/structure/RequestInterpreter.kt]: Also not sure if clients actually _have to_ provide a selection for _all_ types.

Comment (stuebingerb) [reply kgraphql/src/main/kotlin/com/apurebase/kgraphql/schema/structure/RequestInterpreter.kt]: https://graphql.com/learn/interfaces-and-unions/#querying-with-unions does allow queries where not all potential types of a union have a matching fragment but from my understanding that is incorrect: types without fragment result in an empty selection set, and according to the spec, empty selection sets are invalid.

Will keep as-is.

File: kgraphql/src/main/kotlin/com/apurebase/kgraphql/schema/structure/RequestInterpreter.kt (hunk 1 above)
Comment (stuebingerb) [reply kgraphql/src/main/kotlin/com/apurebase/kgraphql/schema/structure/RequestInterpreter.kt]: Adjusted

**Review by mervyn-mccreight — DISMISSED** [review]

**Pull request merged**